sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,602 @@
+through a templated member function:
+
+```
+sycl::specialization_id<int> spec_var;
+
+void func(sycl::queue q) {
+  q.submit([&](sycl::handler &cgh) {
+    cgh.set_specialization_constant<spec_var>(42);
+    cgh.single_task([=](sycl::kernel_handler kh) {
+      int i = kh.get_specialization_constant<spec_var>();
+    });
+  });
+}
+```
+
+Device global variables, by contrast, are referenced by their address:
+
+```
+sycl::ext::oneapi::device_global<int> dev_var;
+
+void func(sycl::queue q) {
+  int val = 42;
+  q.copy(&val, dev_var).wait();      // The 'dev_var' parameter is by reference
+  q.submit([&](sycl::handler &cgh) {
+    cgh.single_task([=] {
+      int i = dev_var;
+    });
+  });
+}
+```
+
+This is a key difference because the compiler does not statically know which
+device global variable is being referenced; we only know the address at
+runtime.  As we will see later, this has a ramification on the integration
+headers and on the mechanism that connects instances of device global variables
+in host code with their corresponding instances in device code.
+
+Another issue relates to the `device_image_scope` property which can be applied
+to a device global variable declaration.  The intent of this property is to
+allow a device global variable to be implemented directly on top of a SPIR-V
+module scope global variable.  When this property is **not** present, an
+instance of a device global variable is shared across all device images that
+are loaded onto a particular device.  By contrast, when this property **is**
+present, each device image has its own instance of the device global variable.
+However, since multiple variable instances have confusing semantics, the API
+requires the user to ensure that each such variable exists in exactly one
+device image.  The extension specification has more details on this property.
+
+The important impact on the design, though, is that device global variables
+declared with the `device_image_scope` property have an implementation that is
+quite different from device global variables that are not declared with this
+property.  The sections below describe both implementations.
+
+
+## Design
+
+### Changes to DPC++ headers
+
+#### Class specializations based on `device_image_scope`
+
+The headers, of course, include the declaration of the new `device_global`
+class, which is described in the [extension specification][1].  The declaration
+of this class uses partial specialization to define the class differently
+depending on whether it has the `device_image_scope` property.  When the
+property is not present, the class has a member variable which is a pointer to
+the underlying type.  Member functions which return a reference to the value
+(e.g. `get`) return the value of this pointer:
+
+```
+template<typename T, typename PropertyListT>
+class device_global {
+  T *usmptr;
+ public:
+  T& get() noexcept { return *usmptr; }
+  /* other member functions */
+};
+```
```

Comment:
in the above, if I instantiate a `device_global` of any type without the `device_image_scope` and then call `myFreshDG.get()` on my newly instantiated variable before calling `queue.copy(someValue, myFreshDG)`, it'll crash because the `usmptr` will not be established, correct?  Is that acceptable?  Or do we want to safely initialize a `val` member in this class as well, and initialize the `usmptr` with its address? 

Is `.get()` usable on any device before `queue.copy()`? Do we care?

Reply:
> it'll crash because the usmptr will not be established, correct?

No, I think this is not correct.  Section "Initializing the device global variables in device code" describes how the `usmptr` field is initialized before any kernel is launched in a module.

> Is .get() usable on any device before queue.copy()? Do we care?

Yes, it is usable, and this is an important use case.  The spec guarantees that the value of a device global variable is zero-initialized, so it is certainly valid for a kernel to access the variable via `.get()` even before any call to `queue::copy()`.  The section "Initializing the device global variables in device code" describes how the USM buffer is zero-initialized, which satisfies this spec requirement.

Reply:
But what about outside of device code?
```
sycl::ext::oneapi::device_global<int> dg_var; //not using device_image_scope, this var will be backed by USM (eventually)
int main(){
   int initialValue = dg_var.get();  //   <--   return *usmptr; !!
   // kernel code below
} 
```

Reply:
Device global variables can only be accessed from device code, not from host code.  The only exceptions are the `copy` and `memcpy` functions defined in the API spec, which copy data between the device and host.  Therefore, the code you show above is not legal.

The API spec addresses this with the following statement:

> It is undefined behavior if the host program directly accesses a device_global or any address obtained from a device_global member function, and similarly it is undefined behavior if a device_global or address obtained on one device from a device_global member function is accessed on a different device. There is no mechanism to obtain addresses of or directly access a device’s device_global allocation within the host program.

Reply:
Thanks. Bringing us around to my original poorly formed question.  By "not legal", do we mean a) if someone writes my above code, and then it segfaults when it's run, that's expected.   or do we mean b) the compiler should not accept it. ?

Reply:
gotcha. it is undefined behavior.  Thanks!!